### ISSUE TITLE
[suggestion] Armour Slots

### ISSUE THREAD

**Issue by SnowyJune678** [MachineMuse/MachineMusePowersuits#831]
The current system of add-ons to the powersuits is set up so you can add all the modules to all the armours with impunity. My idea is that you can only have limited modules per various slot., Each upgrade would have a "isStackable" boolean in the configs which indicates whether the module can be applied to a slot with another module in it. There could also be "Heavy" versions of the suits that allow for two nonstackable modules per slot rather than 1.

Helmet
     Head:  Flight Control(False), Solar Generator(False)
     Eyes: Goggles of Revealing(False), Night Vision(False)
     Mouth: Water Electrolyzer(False), Feeder(False)
     Armour:  Capacitors(False), Basic Plating(False), Advanced Plating(False), Energy Shield(False), Radiation Shielding(True)
Torso/Chestplate
     Front - Extends Armour slot properties: Item Magnet (False), In-Place Assembler (False)
     Back - Extends Armour and Front Torso slot properties: Glider, Parachute, Jetpack
     Armour - Active Camouflage(False), Capacitors(False), Basic Plating(False), Advanced Plating(False), Energy Shield(False), Radiation Shielding(True)
Leggings
     Legs: Swim Boost(True), Sprint Assist(True), Jump assist(True), Stepping assist(True), Static Generator(True)
     Armour:  Capacitors(False), Basic Plating(False), Advanced Plating(False), Energy Shield(False), Radiation Shielding(True)
Boots
     Soles: Shock Absorber(False), Jetboots(False)
     Armour:  Capacitors(False), Basic Plating(False), Advanced Plating(False), Energy Shield(False), Radiation Shielding(True)
Tool
     Palms : Plasma Cannon(False), Blink drive(False), Multimeter(True)
     Backside: Railgun, Retractable Blades (shears)(False), Rototiller (hoe)(False), Pickaxe/Shovel/Axe(False), Diamond Pickaxe Upgrade(True)
     Forearms: Melee Assist(True), Aqua Affinity Upgrade(True)

A config entry could be: 
Plasma Cannon = {
Crafting = {<crafting>} //Crafting Recipe
Slot = "Tool" //Which slot does this go into?
isStackable = False  //Can this be used with other items in the slot?
}

Some other minor requests: Make the tool be off-hand capable, add Advanced Rocketry support, perhaps a chemical protection suit protecting against splash/cloud potions, backpacks, portable ME systems, lasers

**Comment by lehjr** [comment]
This is actually something that is somewhat already in the works. Modules will be craftable items now, with durability instead of weight. The Tinker table will still be needed to install/remove modules, but the code will be different. The number of allowed modules per armor piece will be something that can be set in the config.

**Comment by CTH999** [comment]
Also, there could be a module to allow more modules! That would be nice. (Also, maybe still keep the weight in some form?)

**Comment by lehjr** [comment]
The problem with the weight mechanic is that nothing else in Minecraft has weight, and given that, the weight mechanic can be negated, it's pretty much useless and while adding extra data that has to be tracked and synced. For instance, instead of setting the highest value for the power storage device, the player only needs to carry one in the inventory. I'd rather use a durability mechanic and have modules wear out.

**Comment by CTH999** [comment]
@lehjr true, unless you want to add weight to everything...

or make the things with weight effect you anywhere in the inventory!
Just a thought though!

**Comment by lehjr** [comment]
Yeah, I'm not going to try add weight to every single thing in Minecraft including items from other mods just for a weight mechanic that was meant as a means of balance.

**Comment by SnowyJune678** [comment]
The thing is durability mechanics kind of defeat the purpose of power tools

**Comment by lehjr** [comment]
@0something0 as opposed to the weight mechanic which is easily circumvented, not to mention nothing else in Minecraft has weight?

**Comment by eyeonus** [comment]
No, the weight thing is bad too.

Personally, I don't like the idea of having _either_. Weight is kind of meh, and durability is annoying. I don't like the idea of my railgun breaking in the middle of trying to take down the End Dragon.

Or even worse, my jetpack breaking when I'm 500m in the air....

I don't honestly think an OP-limiter needs to exist at all, but if you really want one, I like the idea of each piece having a certain number of slots, and only being able to install as many modules as there are slots. You could go totally Resident Evil and even have some modules take up multiple slot spaces.


Just remember, the **_only_** other mod (that I'm aware of, anyway) that has upgradable powered armor is DracoEvo, so why not look to that one for inspiration for mitigating OP-Uber-God-Mode?

**Comment by lehjr** [comment]
The durability thing would have been more for low level upgrades, like wood/bamboo armor plating or wooden tools, something that adds a way to create lesser modules with limited resources and fits more with a progression themed gameplay.

It's not that the weight mechanic was a bad idea, as it works well in games like the Fallout series. But in Minecraft, there is no global weight mechanic, so implementing it on just a small selection of items is just out of place, especially when it can be circumvented. On the other hand, the durability mechanic is already deeply entrenched into the game. But, there are other mods that have similar things that do not have a durability, like Mechanism's Atomic Disassembler or Industrialcraft's Quantum suit. And so the argument against both does have valid points. But both of these items do take a good number of resources to create. The Atomic disassembler requires 4 or 5 machines and several diamonds and pieces of obsidian to create, The quantum suit requires several times more resources. 
   
So maybe doing away with the OP limiters on some levels isn't such a bad idea. Implementing a point based system for number of slots a module takes up would actually be far easier to implement than the current system, and as a bonus, would have much less performance impact. The question is whether or not everything should still have adjustable settings or not. Rather than impacting weight, the setting could effect how many slot points the module uses and could be used for heat calculations.

**Comment by eyeonus** [comment]
For most of the modules, I personally don't think they should be adjustable: on literally every single armor module, I always just set it to max and left it there. (Except for the Head module flight assist, which I left at 0 so I could have easy-hover.)

I just had an idea. Get rid of /all/ the sliders, and replace them with all-or-nothings, so either this thing is on at max, or not on at all. For example, the power fist's punch module has two settings, damage and knockback. You can turn damage on but not knockback, which'll do 10 damage, or knockback on without damage, which'll send mobs flying without hurting them, or both on. 

Zap mobs with the plasma ball without any creeper explosion damage to the environment.

Hoe exactly one block of dirt, not a vast 12x12 swath.

For some things, like armor, this on/off toggle doesn't really make sense, so maybe just do that for the power fist modules, and remove the adjustability of armor modules altogether.
